packages/producer-keychain-readmodel-writer/test/consumerServiceV2.test.ts

```diff
@@ -0,0 +1,285 @@
+  toReadModelProducerKeychain,
+  ProducerKeychain,
+  ProducerKeychainKeyAddedV2,
+  ProducerKeychainKeyDeletedV2,
+  ProducerKeychainUserAddedV2,
+  ProducerKeychainUserDeletedV2,
+  EServiceId,
+  ProducerKeychainEServiceAddedV2,
+  ProducerKeychainEServiceRemovedV2,
+  Key,
+} from "pagopa-interop-models";
+import { describe, expect, it } from "vitest";
+import { handleMessageV2 } from "../src/producerKeychainConsumerServiceV2.js";
+import { producerKeychains } from "./utils.js";
+
+describe("Events V2", async () => {
+  const mockProducerKeychain = getMockProducerKeychain();
+  const mockMessage: AuthorizationEventEnvelopeV2 = {
+    event_version: 2,
+    stream_id: mockProducerKeychain.id,
+    version: 1,
+    sequence_num: 1,
+    log_date: new Date(),
+    type: "ProducerKeychainAdded",
+    data: {},
+  };
+
+  it("ProducerKeychainAdded", async () => {
+    const payload: ProducerKeychainAddedV2 = {
+      producerKeychain: toProducerKeychainV2(mockProducerKeychain),
+    };
+
+    const message: AuthorizationEventEnvelopeV2 = {
+      ...mockMessage,
+      type: "ProducerKeychainAdded",
+      data: payload,
+    };
+
+    await handleMessageV2(message, producerKeychains);
+
+    const retrievedProducerKeychain = await producerKeychains.findOne({
+      "data.id": mockProducerKeychain.id,
+    });
+
+    expect(retrievedProducerKeychain?.data).toEqual(
+      toReadModelProducerKeychain(mockProducerKeychain)
+    );
+
+    expect(retrievedProducerKeychain?.metadata).toEqual({
+      version: 1,
+    });
+  });
+
+  it("ProducerKeychainKeyAdded", async () => {
+    await writeInReadmodel(
+      toReadModelProducerKeychain(mockProducerKeychain),
+      producerKeychains,
+      1
+    );
+
+    const key: Key = getMockKey();
+    const updatedProducerKeychain: ProducerKeychain = {
+      ...mockProducerKeychain,
+      keys: [key],
+    };
+    const payload: ProducerKeychainKeyAddedV2 = {
+      producerKeychain: toProducerKeychainV2(updatedProducerKeychain),
+      kid: key.kid,
+    };
+
+    const message: AuthorizationEventEnvelopeV2 = {
+      ...mockMessage,
+      type: "ProducerKeychainKeyAdded",
+      data: payload,
+      version: 2,
+    };
+
+    await handleMessageV2(message, producerKeychains);
+
+    const retrievedProducerKeychain = await producerKeychains.findOne({
+      "data.id": mockProducerKeychain.id,
+    });
+
+    expect(retrievedProducerKeychain?.data).toEqual(
+      toReadModelProducerKeychain(updatedProducerKeychain)
+    );
+    expect(retrievedProducerKeychain?.metadata).toEqual({
+      version: 2,
+    });
+  });
+
+  it("ProducerKeychainKeyDeleted", async () => {
+    const key: Key = getMockKey();
+    const producerKeychain: ProducerKeychain = {
+      ...mockProducerKeychain,
+      keys: [key],
+    };
+    await writeInReadmodel(
+      toReadModelProducerKeychain(producerKeychain),
+      producerKeychains,
+      1
+    );
+
+    const updatedProducerKeychain = mockProducerKeychain;
```

Comment:
Personally I prefer to build the expected data manually, so it's clear what will be the differences between this and the previous version. Same suggestion for other occurrences below.